File: services/agent/src/agent/audit/store.py
"""SQLite 审计存储。与 src-tauri/src/audit/store.rs 共享同一份 schema。

V1.5 (2026-07-31)：audit 表新增 5 列（correlation_id / actor_type / event_type /
task_id / parent_task_id），用于 Phase 12 子 Agent 决策树回放。INSERT 列序必须与
Rust 端 schema.sql 严格镜像（CLAUDE.md §6 红线）。

兼容性：保留旧 `audit(action, payload)` 签名，新增 5 个可选 kwargs；旧调用方零改动。
升级路径：对已存在的旧库执行 ALTER TABLE ADD COLUMN（捕获 "duplicate column" 错）。
"""
from __future__ import annotations

import asyncio
import json
import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

import aiosqlite

from agent.config import settings

logger = logging.getLogger(__name__)

_LOCK = asyncio.Lock()

# V1.5：子 Agent 决策树回放所需 5 列。ALTER TABLE 兼容老库；
# 新建库由 SCHEMA 一次性 CREATE。
_V15_COLUMNS = (
    "correlation_id TEXT",
    "actor_type TEXT",
    "event_type TEXT",
    "task_id TEXT",
    "parent_task_id TEXT",
)


async def audit(
    action: str,
    payload: dict,
    *,
    run_id: str | None = None,
    correlation_id: str | None = None,
    actor_type: Optional[str] = None,
    event_type: Optional[str] = None,
    task_id: Optional[str] = None,
    parent_task_id: Optional[str] = None,
    db_path: str | None = None,
) -> None:
    """Append an entry to the audit log. Fire-and-forget, but durable.

    V1.5 新增 5 个 kwargs：
      - correlation_id: 同一棵决策树的所有事件共享此 ID
      - actor_type: 'user' / 'main_agent' / 'sub_agent' / 'system'
      - event_type: 结构化事件名（与 action 平行，向后兼容）
      - task_id: 子 Agent sub_agent_id
      - parent_task_id: 父 sub_agent_id

    `db_path` 优先（默认 None 时用 settings.audit_db_path）。
    显式传 db_path 可避免：pytest async 测试间 monkeypatch.chdir 不稳定时，
    写入路径与测试期望路径错位。
    """
    target = db_path or settings.audit_db_path
    Path(target).parent.mkdir(parents=True, exist_ok=True)
    async with _LOCK, aiosqlite.connect(target) as db:
        # 1) 先建表（IF NOT EXISTS → 旧库跳过）
        await db.executescript(SCHEMA_CREATE_TABLE)
        # 2) 旧库 ALTER TABLE ADD COLUMN（新库已包含则跳过）
        await _ensure_v15_columns(db)
        # 3) 最后建索引（此时 correlation_id 列一定存在）
        await db.executescript(SCHEMA_INDEXES)
        await db.execute(
            "INSERT INTO audit("
            "  action, payload, ts, run_id,"
            "  correlation_id, actor_type, event_type, task_id, parent_task_id"
            ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
            (
                action,
                json.dumps(payload, ensure_ascii=False, default=str),
                datetime.now(timezone.utc).isoformat(),
                run_id,
                correlation_id,
                actor_type,
                event_type,
                task_id,
                parent_task_id,
            ),
        )
        await db.commit()


async def _ensure_v15_columns(db: aiosqlite.Connection) -> None:
    """对已存在的旧库加 5 列（不存在则跳过）。幂等。"""
    for col_def in _V15_COLUMNS:
        col_name = col_def.split()[0]
        try:
            await db.execute(f"ALTER TABLE audit ADD COLUMN {col_def}")
        except Exception as exc:  # noqa: BLE001
            # "duplicate column name" → 已存在，忽略
            msg = str(exc).lower()
            if "duplicate column" not in msg:
                logger.warning("ALTER TABLE audit ADD COLUMN %s 失败: %s", col_name, exc)


async def search(query: str, limit: int = 200) -> list[dict]:
    async with aiosqlite.connect(settings.audit_db_path) as db:
        cur = await db.execute(
            "SELECT action, payload, ts, run_id FROM audit "
            "WHERE action LIKE ? OR payload LIKE ? ORDER BY id DESC LIMIT ?",
            (f"%{query}%", f"%{query}%", limit),
        )
        rows = await cur.fetchall()
    return [
        {"action": a, "payload": json.loads(p), "ts": t, "run_id": r}
        for (a, p, t, r) in rows
    ]


async def search_by_correlation(
    correlation_id: str,
    limit: int = 500,
    db_path: str | None = None,
) -> list[dict]:
    """V1.5 新增：按 correlation_id 回放一整棵决策树。

    返回按时间升序的事件列表（先发生先返回），便于上溯决策树。
    """
    target = db_path or settings.audit_db_path
    async with aiosqlite.connect(target) as db:
        cur = await db.execute(
            "SELECT action, payload, ts, run_id, correlation_id, actor_type, "
            "       event_type, task_id, parent_task_id "
            "FROM audit WHERE correlation_id = ? "
            "ORDER BY id ASC LIMIT ?",
            (correlation_id, limit),
        )
        rows = await cur.fetchall()
    return [
        {
            "action": a,
            "payload": json.loads(p),
            "ts": t,
            "run_id": r,
            "correlation_id": c,
            "actor_type": ac,
            "event_type": ev,
            "task_id": tk,
            "parent_task_id": pt,
        }
        for (a, p, t, r, c, ac, ev, tk, pt) in rows
    ]


SCHEMA_CREATE_TABLE = """
CREATE TABLE IF NOT EXISTS audit (
    id              INTEGER PRIMARY KEY AUTOINCREMENT,
    action          TEXT NOT NULL,
    payload         TEXT NOT NULL,
    ts              TEXT NOT NULL,
    operator        TEXT,
    run_id          TEXT,
    correlation_id  TEXT,
    actor_type      TEXT,
    event_type      TEXT,
    task_id         TEXT,
    parent_task_id  TEXT
);
"""

# 索引必须在 _ensure_v15_columns 之后执行，否则旧库缺少 correlation_id 列会报错
SCHEMA_INDEXES = """
CREATE INDEX IF NOT EXISTS idx_audit_action      ON audit(action);
CREATE INDEX IF NOT EXISTS idx_audit_ts          ON audit(ts);
CREATE INDEX IF NOT EXISTS idx_audit_run         ON audit(run_id);
CREATE INDEX IF NOT EXISTS idx_audit_correlation ON audit(correlation_id);
CREATE INDEX IF NOT EXISTS idx_audit_actor_event ON audit(actor_type, event_type);
CREATE INDEX IF NOT EXISTS idx_audit_task        ON audit(task_id);
"""

File: apps/desktop/src-tauri/src/audit/store.rs
//! SQLite-backed audit log. Schema kept in sync with the Python Agent's
//! `services/agent/src/agent/audit/schema.sql`.
//!
//! V1.5 (2026-07-31): 加 5 列（correlation_id / actor_type / event_type /
//! task_id / parent_task_id），用于 Phase 12 子 Agent 决策树回放。Python +
//! Rust 双 schema INSERT 列序必须严格镜像（CLAUDE.md §6 红线）。

use std::path::Path;

use chrono::Utc;
use rusqlite::{params, Connection};

use crate::error::{AppError, AppResult};

pub struct AuditStore {
    conn: Mutex<Connection>,
}

use std::sync::Mutex;

impl AuditStore {
    pub fn open(path: &Path) -> AppResult<Self> {
        // rusqlite::Connection::open 不会自动建父目录，
        // 全新安装时 `%APPDATA%\eaide\` 不存在会报错，导致整个 setup() 失败、窗口不弹出。
        if let Some(parent) = path.parent() {
            if !parent.as_os_str().is_empty() {
                std::fs::create_dir_all(parent)?;
            }
        }
        let conn = Connection::open(path)?;
        // 启用 WAL 模式 —— 允许 Rust 和 Python 进程同时读写同一个数据库
        // 默认的 rollback journal 模式在跨进程并发时会导致频繁 SQLITE_BUSY
        conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA busy_timeout=5000;")?;
        // 应用共享 schema（幂等 CREATE IF NOT EXISTS）
        conn.execute_batch(SCHEMA)?;
        // V1.5：老库 ALTER TABLE ADD COLUMN（CREATE IF NOT EXISTS 已包含新列时跳过）
        for col_def in V15_COLUMNS {
            let col_name = col_def.split_whitespace().next().unwrap_or("");
            let sql = format!("ALTER TABLE audit ADD COLUMN {} DEFAULT NULL", col_def);
            if let Err(e) = conn.execute_batch(&sql) {
                // "duplicate column name" 表示列已存在，忽略；其他错则记录但继续
                let msg = e.to_string().to_lowercase();
                if !msg.contains("duplicate column") {
                    eprintln!(
                        "[audit] ALTER TABLE audit ADD COLUMN {} 失败（旧库迁移跳过）: {}",
                        col_name, e
                    );
                }
            }
        }
        Ok(Self { conn: Mutex::new(conn) })
    }

    /// 向后兼容的旧 API —— 仅写 action / payload / ts / run_id 4 列。
    /// V1.5 之前调用方零改动；新调用方请使用 `append_v15`。
    pub fn append(&self, action: &str, payload: serde_json::Value) -> AppResult<()> {
        let conn = self
            .conn
            .lock()
            .map_err(|e| AppError::Internal(format!("审计数据库锁 poisoned（前次 panic 导致）: {}", e)))?;
        conn.execute(
            "INSERT INTO audit(action, payload, ts) VALUES (?, ?, ?)",
            params![action, payload.to_string(), Utc::now().to_rfc3339()],
        )?;
        Ok(())
    }

    /// V1.5 新增：写入全 9 列（与 Python audit() INSERT 列序严格镜像）。
    ///
    /// 列序：action / payload / ts / run_id / correlation_id / actor_type /
    ///       event_type / task_id / parent_task_id
    #[allow(clippy::too_many_arguments)]
    pub fn append_v15(
        &self,
        action: &str,
        payload: serde_json::Value,
        run_id: Option<&str>,
        correlation_id: Option<&str>,
        actor_type: Option<&str>,
        event_type: Option<&str>,
        task_id: Option<&str>,
        parent_task_id: Option<&str>,
    ) -> AppResult<()> {
        let conn = self
            .conn
            .lock()
            .map_err(|e| AppError::Internal(format!("审计数据库锁 poisoned（前次 panic 导致）: {}", e)))?;
        conn.execute(
            concat!(
                "INSERT INTO audit(",
                "  action, payload, ts, run_id,",
                "  correlation_id, actor_type, event_type, task_id, parent_task_id",
                ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
            ),
            params![
                action,
                payload.to_string(),
                Utc::now().to_rfc3339(),
                run_id,
                correlation_id,
                actor_type,
                event_type,
                task_id,
                parent_task_id,
            ],
        )?;
        Ok(())
    }

    pub fn search(&self, query: &str, limit: u32) -> AppResult<Vec<serde_json::Value>> {
        let conn = self
            .conn
            .lock()
            .map_err(|e| AppError::Internal(format!("审计数据库锁 poisoned（前次 panic 导致）: {}", e)))?;
        let mut stmt = conn.prepare(
            "SELECT action, payload, ts FROM audit
             WHERE action LIKE ?1 OR payload LIKE ?1
             ORDER BY id DESC LIMIT ?2",
        )?;
        let pat = format!("%{}%", query);
        let rows = stmt.query_map(params![pat, limit], |row| {
            let action: String = row.get(0)?;
            let payload: String = row.get(1)?;
            let ts: String = row.get(2)?;
            Ok(serde_json::json!({
                "action": action,
                "payload": serde_json::from_str::<serde_json::Value>(&payload)
                    .unwrap_or(serde_json::Value::String(payload)),
                "ts": ts,
            }))
        })?;
        let mut out = Vec::new();
        for r in rows { out.push(r?); }
        Ok(out)
    }

    /// V1.5 新增：按 correlation_id 回放一整棵决策树。
    pub fn search_by_correlation(
        &self,
        correlation_id: &str,
        limit: u32,
    ) -> AppResult<Vec<serde_json::Value>> {
        let conn = self
            .conn
            .lock()
            .map_err(|e| AppError::Internal(format!("审计数据库锁 poisoned（前次 panic 导致）: {}", e)))?;
        let mut stmt = conn.prepare(
            concat!(
                "SELECT action, payload, ts, run_id, correlation_id, actor_type, ",
                "       event_type, task_id, parent_task_id ",
                "FROM audit WHERE correlation_id = ?1 ",
                "ORDER BY id ASC LIMIT ?2",
            ),
        )?;
        let rows = stmt.query_map(params![correlation_id, limit], |row| {
            let action: String = row.get(0)?;
            let payload: String = row.get(1)?;
            let ts: String = row.get(2)?;
            let run_id: Option<String> = row.get(3)?;
            let correlation_id: Option<String> = row.get(4)?;
            let actor_type: Option<String> = row.get(5)?;
            let event_type: Option<String> = row.get(6)?;
            let task_id: Option<String> = row.get(7)?;
            let parent_task_id: Option<String> = row.get(8)?;
            Ok(serde_json::json!({
                "action": action,
                "payload": serde_json::from_str::<serde_json::Value>(&payload)
                    .unwrap_or(serde_json::Value::String(payload)),
                "ts": ts,
                "run_id": run_id,
                "correlation_id": correlation_id,
                "actor_type": actor_type,
                "event_type": event_type,
                "task_id": task_id,
                "parent_task_id": parent_task_id,
            }))
        })?;
        let mut out = Vec::new();
        for r in rows { out.push(r?); }
        Ok(out)
    }
}

/// V1.5 新增 5 列的 ALTER TABLE 定义。列序与 schema.sql 镜像。
const V15_COLUMNS: &[&str] = &[
    "correlation_id TEXT",
    "actor_type TEXT",
    "event_type TEXT",
    "task_id TEXT",
    "parent_task_id TEXT",
];

const SCHEMA: &str = include_str!("schema.sql");

impl AuditStore {
    /// 内存版空 AuditStore —— 真实 DB 初始化失败时的兜底，所有方法可用，只是写入会丢失。
    pub fn empty() -> AppResult<Self> {
        let conn = Connection::open_in_memory()?;
        conn.execute_batch("PRAGMA journal_mode=WAL; PRAGMA busy_timeout=5000;")?;
        conn.execute_batch(SCHEMA)?;
        // 内存版也执行 V15 迁移（虽无意义，保持一致性）
        for col_def in V15_COLUMNS {
            let sql = format!("ALTER TABLE audit ADD COLUMN {} DEFAULT NULL", col_def);
            let _ = conn.execute_batch(&sql);
        }
        Ok(Self { conn: Mutex::new(conn) })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// V1.5：验证 append_v15 写入 9 列无误 + search_by_correlation 能按 correlation_id 拉回。
    #[test]
    fn audit_v15_roundtrip() {
        let store = AuditStore::empty().unwrap();
        let payload = serde_json::json!({"sub_agent_id": "sub-1"});
        store
            .append_v15(
                "SUB_AGENT_SPAWN",
                payload,
                Some("run-1"),
                Some("corr-1"),
                Some("sub_agent"),
                Some("spawn"),
                Some("sub-1"),
                None,
            )
            .unwrap();
        store
            .append_v15(
                "SUB_AGENT_DONE",
                serde_json::json!({"status": "ok"}),
                Some("run-1"),
                Some("corr-1"),
                Some("sub_agent"),
                Some("done"),
                Some("sub-1"),
                None,
            )
            .unwrap();
        let rows = store.search_by_correlation("corr-1", 10).unwrap();
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0]["action"], "SUB_AGENT_SPAWN");
        assert_eq!(rows[0]["task_id"], "sub-1");
        assert_eq!(rows[0]["actor_type"], "sub_agent");
        assert_eq!(rows[1]["action"], "SUB_AGENT_DONE");
    }

    /// V1.5：旧 API `append` 仍然只写 3 列（向后兼容）
    #[test]
    fn audit_legacy_append_still_works() {
        let store = AuditStore::empty().unwrap();
        store
            .append("OLD_ACTION", serde_json::json!({"k": "v"}))
            .unwrap();
        let rows = store.search("OLD_ACTION", 10).unwrap();
        assert_eq!(rows.len(), 1);
        assert_eq!(rows[0]["action"], "OLD_ACTION");
    }
}
